File: Project/Simulator/Organisms/Animals/CyberSheep.java
package Simulator.Organisms.Animals;

import org.json.JSONObject;

import Simulator.World;
import Simulator.Tile;
import Simulator.Organisms.Organism;
import Utils.RNG;

import java.util.ArrayList;
import java.util.List;

public class CyberSheep extends Animal {
    public CyberSheep(World world) {
        super(11, 4, world, Type.CYBER_SHEEP);
    }

    public CyberSheep(JSONObject json, World world) {
        super(json, world);
    }

    @Override
    public Animal construct() {
        return new CyberSheep(world);
    }

    @Override
    public void action() {
        int index = seekClosestSosnowskyHogweed();
        if (index == -1) {
            super.action();
            return;
        }
        Tile target = nextTileToSosnowskyHogweed(index);
        move(target);
        // if (target == null) {
        // super.action();
        // return;
        // }
        // move(target);
    }

    private int seekClosestSosnowskyHogweed() {
        int index = -1;
        int min_distance = -1;
        for (Organism organism : world.getOrganisms()) {
            if (organism.getType() == Type.SOSNOWSKY_HOGWEED) {
                int distance = world.getDistance(tile, organism.getTile());
                if (min_distance == -1 || distance < min_distance) {
                    min_distance = distance;
                    index = world.getOrganisms().indexOf(organism);
                }
            }
        }
        return index;
    }

    private Tile nextTileToSosnowskyHogweed(int index) {
        Tile target = world.getOrganism(index).getTile();
        int distance = world.getDistance(tile, target);
        List<Tile> neighbours = tile.getNeighbours();
        List<Tile> valid = new ArrayList<>();
        for (Tile neighbour : neighbours) {
            if (neighbour == null) {
                continue;
            }
            if (world.getDistance(neighbour, target) < distance) {
                valid.add(neighbour);
            }
        }
        if (valid.isEmpty()) {
            return null;
        }
        target = valid.get(RNG.roll(0, valid.size() - 1));

        return target;
    }
}

File: Project/Simulator/Organisms/Animals/CyberSheep.py
import re
from tracemalloc import start
from .Animal import Animal
from Simulator.Organisms.Organism import Type
from Simulator.Organisms.Plants.SosnowskyHogweed import SosnowskyHogweed

from collections import deque


class CyberSheep(Animal):
    def __init__(self, world, json=None):
        super().__init__(11, 4, world, Type.CYBER_SHEEP)

    def construct(self):
        return CyberSheep(self._world)

    def pathfind(self):
        queue = deque()
        queue.append(self._tile)
        visited = set([self._tile])
        found = None
        while queue:
            current = queue.popleft()
            if isinstance(current.getOrganism(), SosnowskyHogweed):
                found = current
                break
            for neighbour in current.getNeighbours():
                if neighbour in visited:
                    continue
                visited.add(neighbour)
                queue.append(neighbour)
                neighbour.setParent(current)
        if found is None:
            return None
        queue.clear()
        queue.append(current)
        visited.remove(current)
        while queue:
            current = queue.popleft()
            if (current.getParent() is self._tile):
                return current
            for neighbour in current.getNeighbours():
                if neighbour not in visited:
                    continue
                visited.remove(neighbour)
                queue.append(neighbour)
        raise Exception("Error of pathfinding")

    def action(self):
        target = None
        if any(isinstance(instance, SosnowskyHogweed) for instance in self._world.getOrganisms()):
            target = self.pathfind()
        if target is None:
            super().action()
            return
        self.move(target)
